scripts/validate_response.py

```diff
@@ -0,0 +1,150 @@
+        ans_vec = self._embedding_model.get_text_embedding(answer)
+
+        # Distance score
+        cos_score = cosine(res_vec, ans_vec)
+        euc_score = euclidean(res_vec, ans_vec)
+
+        # Naive length consideration with reduced weightage.
+        len_res, len_ans = len(response), len(answer)
+        len_score = (abs(len_res - len_ans) / (len_res + len_ans)) * 0.1
+
+        score = len_score + (cos_score + euc_score) / 2
+        # TODO: Consider both contextual/non-contextual embedding.
+
+        print(
+            f"cos_score: {cos_score}, "
+            f"euc_score: {euc_score}, "
+            f"len_score: {len_score}\n"
+            f"final_score: {score}"
+        )
+        return score
+
+    def get_response_quality(self, args, qa_pairs):
+        """Get response quality."""
+        result_dict = defaultdict(list)
+
+        for idx in range(min(len(qa_pairs), args.num_questions)):
+            question = qa_pairs[idx]["question"]
+            answer = qa_pairs[idx]["answer"]
+
+            response = requests.post(
+                # API question validator can be disabled.
+                "http://localhost:8080/v1/query",
+                json={"query": question},
+                timeout=90,
+            ).json()["response"]
+
+            if args.scenario == "without_rag":
+                response = response.replace(
+                    "The following response was generated without access to reference content:\n\n",
+                    "",
+                )
+
+            print(f"Calculating score for query: {question}")
+            score = self.get_similarity_score(response, answer)
+
+            if score > args.threshold:
+                print(f"Response is not as expected for question: {question}")
+
+            result_dict["question"].append(question)
+            result_dict["answer"].append(answer)
+            result_dict["llm_response"].append(response)
+            result_dict["recent_score"].append(score)
+
+        result_df = DataFrame.from_dict(result_dict)
+        result_df["threshold"] = args.threshold
+        return result_df
+
+
+def main():
+    """Validate LLM response."""
+    args = _args_parser(sys.argv[1:])
+    print(f"Arguments passed: {args}")
+
+    parent_dir = os.path.dirname(os.path.dirname(__file__)) + "/tests"
+
+    with open(f"{parent_dir}/test_data/question_answer_pair.json") as qna_f:
+        qa_pairs = json.load(qna_f)
+        qa_pairs = qa_pairs.get(args.model, {}).get(args.scenario, [])
+
+    result_df = ResponseValidation().get_response_quality(args, qa_pairs)
+
+    if len(result_df) > 0:
+        result_dir = f"{parent_dir}/test_results"
+        os.makedirs(result_dir, exist_ok=True)
+        result_file = (
+            f"{result_dir}/question_answer_result_{args.scenario}_{args.model}.csv"
+        )
+        result_df.to_csv(result_file, index=False)
+        print(f"Result is saved to {result_file}")
```

Comment:
What's the idea behind persisting these results in a file. Why not just log them in whatever format we want? Because that file will be created in PROW environment during runtime and will be destroyed after the test ends.

Reply:
Understood.